### PR TITLE
bpo-47142: Refine Traversable.

### PR DESCRIPTION
**Pull request by jaraco**
Addresses concerns reported in [bpo-47142](https://bugs.python.org/issue47142).

### CODE REVIEW

**Comment by jaraco** [comment]
@encukou, please take a look and let me know what you think.

**Review by FFY00 — APPROVED** [review]
This looks alright to me FWIW.

File: importlib_resources/simple.py [hunk 1]
```diff
@@ -99,10 +99,13 @@ def iterdir(self):
...
 
-    def joinpath(self, name):
+    def joinpath(self, *names):
+        if not names:
+            return self
+        name, rest = names[0], names[1:]
         return next(
             traversable for traversable in self.iterdir() if traversable.name == name
-        )
+        ).joinpath(*rest)
```
[the remark this answers is not in the document]
Comment (jaraco) [reply importlib_resources/simple.py:108]: This whole module is untested and potentially unused. I've presented it as a proof of concept, but I'm not aware of any consumers of it. I'm not inclined to spend a lot of time refining the implementation if no one is getting value from it.

That said, you're right, it really should honor `/`-separated names.

We could consider moving it to `Traversable`. I'm a little conflicted on whether this additional detail should be part of the more visible `Traversable` or reserved with `simple` where it's mainly/solely used.

[the remark this answers is not in the document]
Comment (jaraco) [reply importlib_resources/simple.py:108]: See #249

[the remark this answers is not in the document]
Comment (encukou) [reply importlib_resources/simple.py:108]: Hmm. `importlib_resources.readers.MultiplexedPath` seems to have the same issue. Moving the intended logic to `Traversable` could help there as well.
`MultiplexedPath` also seems to try creating paths non-existing resources, is that expected?

[the remark this answers is not in the document]
Comment (FFY00) [reply importlib_resources/simple.py:108]: Yes, it mimics `pathlib.Path` in that detail.

Currently, whether an exception is raised or not is implementation defined.

https://github.com/python/cpython/pull/26272 is slightly relevant here.

[the remark this answers is not in the document]
Comment (jaraco) [reply importlib_resources/simple.py:108]: > Hmm. `importlib_resources.readers.MultiplexedPath` seems to have the same issue. Moving the intended logic to `Traversable` could help there as well. `MultiplexedPath` also seems to try creating paths non-existing resources, is that expected?

See #250

**Pull request merged**